mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java

```diff
@@ -113,6 +114,67 @@ private static UidCriterion findConjugatedUidCriterion(List<Criterion> crits) {
         return null;
     }
     
+    /**
+     * Walks down the query tree's conjunctions to find the highest necessary mail fetch type.
+     * @param crits - list of Criterion to search from
+     * @return required fetch type - metadata, headers, or full
+     */
+    private static FetchType getFetchTypeForCriteria(List<Criterion> crits) {
+        FetchType maximum = FetchType.METADATA;
+        for (Criterion crit : crits) {
+            if (crit instanceof ConjunctionCriterion) {
+                return getFetchTypeForCriteria(((ConjunctionCriterion) crit)
+                    .getCriteria());
+            } else {
+                maximum = maxFetchType(maximum, getFetchTypeForCriterion(crit));
+            }
+        }
+        return maximum;
+    }
+
+    private static FetchType getFetchTypeForCriterion(Criterion crit) {
+        if (crit instanceof SearchQuery.AllCriterion || crit instanceof SearchQuery.TextCriterion) {
+            return FetchType.FULL;
+        } else
+        if (crit instanceof SearchQuery.HeaderCriterion || crit instanceof SearchQuery.MimeMessageIDCriterion) {
+            return FetchType.HEADERS;
+        } else {
+            return FetchType.METADATA;
+        }
+    }

```

Comment:
If there is a conjunction criteria, we still need to take into account the next criterias.

I think here is a very good example where functional programing might be more efficient? (map - reduce)

For instance consider `AND(CritA, CritB), CritC` then with your current code CritC is ignored. Which is a bug.


```suggestion
    private static FetchType getFetchTypeForCriteria(List<Criterion> crits) {
        return crits.stream()
            .map(crit -> getFetchTypeForCriterion(crit))
            .reduce((a, b) -> maxFetchType(a, b))
            .orElse(FetchType.METADATA);
    }

    private static FetchType getFetchTypeForCriterion(Criterion crit) {
        if (crit instanceof ConjunctionCriterion) {
            return getFetchTypeForCriteria(((ConjunctionCriterion) crit)
                    .getCriteria())
        }
        if (crit instanceof SearchQuery.AllCriterion || crit instanceof SearchQuery.TextCriterion) {
            return FetchType.FULL;
        }
        if (crit instanceof SearchQuery.HeaderCriterion || crit instanceof SearchQuery.MimeMessageIDCriterion) {
            return FetchType.HEADERS;
        } 
         return FetchType.METADATA;
    }

```

Reply:
Oops. I originally copied it from findConjugatedUidCriterion, but then envolved it from abbreviated to complete traversal and forgot to adapt it accordingly. Nice catch!